### PR TITLE
PCHR-1421: Loading comments from API

### PR DESCRIPTION
**Pull request by deb1990**
As part of this PR(and https://github.com/compucorp/civihr-employee-portal-theme/pull/144) comments support has been added to `LeaveRequestModal`

Screengrab:
Addition and removal of comments: http://recordit.co/lopULZS3uG
Manager can remove all comments: http://recordit.co/lUfjylyhVk

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 1]
```diff
@@ -825,6 +902,47 @@ define([
...
+       *
+       * @return {Promise}
+       */
+      function loadContactNames () {
+        var contactIDs = [],
+          self = this;
+
+        _.each(self.request.comments, function (comment) {
+          //Push only unique contactId's which are not same as logged in user
+          if(comment.contact_id != self.directiveOptions.contactId && contactIDs.indexOf(comment.contact_id) === -1) {
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: will self.request.contat_id do?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 2]
```diff
@@ -825,6 +902,47 @@ define([
...
+          }
+        });
+
+        return Contact.all({
+          id: { IN: contactIDs }
+        },{
+          page: 1,
+          size: 0
+        }).then(function (contacts) {
+          self.comments.contacts = _.indexBy(contacts.list, 'contact_id');
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: why maintain a separate comments array when one is already available on leave request instance?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 3]
```diff
@@ -88,7 +95,8 @@ define([
          */
         defaultCustomData: function () {
           return {
-            comments: []
+            comments: [],
+            commentsToBeDeleted: []
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: can we have attribute in comments array itself that will mark it for deletion like `canDelete`?

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 4]
```diff
@@ -93,11 +97,24 @@ define([
...
+          }
         }
       };
 
       /**
+       * Add a comment into comments array, also clears the comments textbox
+       */
+      this.addComment = function () {
+        this.request.comments.push({
+          contact_id: this.directiveOptions.contactId,
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: No here the logged in users contact ID should be used. The leave request might be someone else's but the comment should be added for the currently logged in user.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 5]
```diff
@@ -175,6 +192,29 @@ define([
...
+      };
+
+      /**
+       * Returns the comment author name
+       * @param {String} contact_id
+       *
+       * @return {String}
+       */
+      this.getCommentorName = function (contact_id) {
+        if(contact_id == this.directiveOptions.contactId) {
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: No here the logged in users contact ID should be used. The leave request might be someone else's but the comment should be added for the currently logged in user.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js (hunk 2 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Here, `comments` is not an array, its an object, which is a place holder for the comment box text model and also the contact names for all comment authors.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js (hunk 1 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: `//Push only unique contactId's which are not same as logged in user`
Logic is based on logged in user, and the leave request can also be someone else's. So it wont work.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 6]
```diff
@@ -75,7 +79,7 @@ define([
         isChangeExpanded: false,
         multipleDays: true,
         workedDate: null,
-        userDateFormat: HR_settings.DATE_FORMAT,
+        userDateFormat: HR_settings.DATE_FORMAT.toUpperCase(),
```
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: Its working.

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js [hunk 7]
```diff
@@ -46,6 +46,10 @@ define([
           breakdown: []
         }
       };
+      this.comments = {
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/request-ctrl.js]: this is for a single comment so can it be `this.comment`

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js [hunk 8]
```diff
@@ -88,7 +91,8 @@ define([
...
```
Comment (dhavalsharma) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/models/instances/leave-request-instance.js]: `commentsToBeDeleted` has to be removed from toAPI also?

**Pull request merged**